CBPs/cbp-0002.md

```diff
@@ -0,0 +1,632 @@
+
+2. ERC667
+    - adds transferAndCall which combines transfer function and call function, and solve the problems solved by ERC223
+
+3. ERC777
+    - use send function to repalce transfer function
+    - use the hook function to controls whether to accept transfer
+    - use the operator authorization function to replace the approve function as a proxy transfer solution
+
+#### Improved standards
+
+```
+service: {
+  name: () -> (text) query;
+  symbol: () -> (text) query;
+  decimals: () -> (nat8) query;
+  totalSupply: () -> (nat) query;
+
+  balanceOf: (owner: principal) -> (nat) query;
+  allowance: (owner: principal, spender: principal) -> (nat) query;
+  approve: (spender: principal, value: nat) -> (bool);
+  transferFrom: (sender: principal, receiver: principal, value: nat) -> (bool);
+  transfer: (receiver: principal, value: nat, args:opt vec nat8) -> (bool);
+}
+```
+
+## Suitable for Dfinity
+
+### Problems to be solved
+
+The design of Token Standard will consider the difference between Dfinity and Ethereum fully, and clarify the problems to be solved:
+
+1. No atomicity similar to EVM cross-contract calls
+
+- Solution: [sagas](https://www.cs.cornell.edu/andru/cs711/2002fa/reading/sagas.pdf) is a good solution
+
+2. No built-in event-emit support similar to EVM
+
+- Question: 
+  - How do others know that the transaction happened？
+  - How to check transaction history？
+
+- Consideration: There are two ideas (Pubsub/Notify) for question 1 on the forum
+
+  - When the Token is transferred, Notify informs the recipient, which can fill the missing EVENT. But When the Token recipient not a canister, which means can not notify, it is necessary to support query transaction history.
+
+  - Token does not have sufficient reason to implement Pub/Sub to release transaction information to a third party that is not the recipient
+
+- Solution: 
+  - Notify is a better way; 
+  - It's necessary to support query transaction history;
+
+3. Built-in storage support
+
+- Question: The current storage increase to 8G, but the memory storage is still 4G
+
+- Consideration:
+
+  - Tx history should support query
+
+  - Built-in storage can support Token to store more self-describing information
+
+- Solution:
+
+  - Auto-scaling storage solution
+  - Token implements self-description
+
+4. Reverse gas model
+ - Question: How to prevent ddos attacks
+ - Consideration: The call of the canister does not require the caller to pay gas fees(canister pay the gas fees)
+
+- Solution: Any update call to canister should be charged to the caller
+
+5. There are two different identities in Dfinity, Internet Identity (II for short) and Principal ID
+
+- Question: Which identity to use as a token standard user？
+
+- Consideration: Dfinity's II is an implementation of DID, and II is based on Principal ID
```

Comment:
Doesn't II look like a Principal ID to canisters?
Should a token standard care too much about what kind of client is communicating with it?

Reply:
Principal & AccountId somehow makes sense can be confusing, but as a token standard, choosing to support only Principal is the easiest, but being AccountID compatible will keep users from passively choosing

Reply:
Oh yeah, AccountIDs aren't great. Hash(PrincipalID, SubAccount) is equivalent to AccountID.
I think (PrincipalID, SubAccount) may make more sense, or just drop the SubAccounts

Reply:
Here my idea is different, I think IC official team design AccountID is in privacy protection considerations; (PrincipalID, SubAccount) lead to different sub-accounts, will expose Principal. but after AccountID, can play a privacy protection role, so the token standard should not choose to reduce in privacy protection level, which is why I think it is necessary to support AccountID

Reply:
You are right, but I think most people are lazy, and also creating more private keys can be prone to loss or forgetting, if a Principal can serve the purpose of protecting privacy is a better and simpler way